server/server.go

```diff
@@ -36,19 +52,101 @@ func upload(w http.ResponseWriter, r *http.Request) {
 	w.Write(responseS)
 }
 
-// Reads the compressed file and invokes function to decompress it
+// Builds and runs the assignment uploaded.
+func buildRun(w http.ResponseWriter, r *http.Request) {
+
+	fmt.Println("Build and Run Endpoint Hit")
+	w.Header().Set("Access-Control-Allow-Origin", "*")
+
+	var outputString string
+	var currDir string
+	var err error
+
+	outputString, currDir = navigateToWorkDir()
+	if outputString == "" {
+		outputString, err = runCommand(AssignmentData.CommandToCompile)
+
+		if err == nil {
+			finalCmd := AssignmentData.CommandToExecute
+			for _, value := range AssignmentData.CmdlineArgs {
+				finalCmd = fmt.Sprintf("%s %s", finalCmd, value)
+			}
+			fmt.Println(finalCmd)
+			outputString, err = runCommand(finalCmd)
+		}
+
+		err = os.Chdir(currDir)
+		if err != nil {
+			log.Fatalf("error while navigating to the current directory: %v", err)
+		}
+	}
+
+	response, err := json.Marshal(outputString)
+	if err != nil {
+		log.Println(err)
+	}
+	// Write the response to be sent to client.
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	w.Write(response)
+}
+
+func navigateToWorkDir() (string, string) {
+	var responseS = ""
+	workDir := filepath.Join(AssignmentData.RootDir, AssignmentData.WorkDir)
+	currDir, _ := os.Getwd()
+	err := os.Chdir(filepath.Join(currDir, constants.AssignmentsDir, workDir))
+	if err != nil {
+		responseS = "Error while navigating to working directory"
+		log.Println("error while navigating to the working directory: ", err)
+	}
+	return responseS, currDir
+}
+
+// Runs the provided command.
+func runCommand(cmdStr string) (string, error) {
+	var out bytes.Buffer
+	var stderr bytes.Buffer
+	var output string
+	cmd := exec.Command("/bin/sh", "-c", cmdStr)
+	cmd.Stdout = &out
+	cmd.Stderr = &stderr
+	err := cmd.Run()
+	if err != nil {
+		output = fmt.Sprintf("%v", stderr.String())
+		return output, err
+	}
+
+	output = fmt.Sprintf("%v", out.String())
+	return output, nil
+}
+
+// Reads the compressed file and invokes function to decompress it.
 func readFormData(r *http.Request) string {
 	fileHeader := make([]byte, 512)
 
 	// Returns the first file for the given key 'file'.
 	file, handler, err := r.FormFile(constants.FormFileKey)
 	if err != nil {
 		responseString := `"File Error":"Error in retrieving the file"`
-		log.Println("Error Retrieving the File", err)
+		log.Println("error Retrieving the file", err)
 		return responseString
 	}
 
-	// TODO: read the command line arguments
+	AssignmentData.CmdlineArgs = make(map[string]string)
+	for index := 1; index <= len(r.Form); index++ {
+		keyName := fmt.Sprintf("%s%d", constants.CmdArgKeyName, index)
+		key := r.FormValue(keyName)
+
+		argName := fmt.Sprintf("%s%d", constants.CmdArgValueName, index)
+		arg := r.FormValue(argName)
+		AssignmentData.CmdlineArgs[key] = arg
+	}
+
+	// Read the working directory and command to run.
```

Comment:
comment doesn't mention anything about reading in the command to compile the assignment.

Reply:
Added